File: src/greedy/CoinZero.java
package greedy;

import java.io.*;
import java.util.ArrayList;
import java.util.List;

public class CoinZero {
    public static void main(String[] args) throws IOException {
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] NM = br.readLine().split(" ");
        int N = Integer.parseInt(NM[0]), M = Integer.parseInt(NM[1]);

        List<Integer> coins = new ArrayList<>();

        for (int i = 0; i < N; i++) {
            int coin = Integer.parseInt(br.readLine());

            if(coin <= M){
                coins.add(coin);
            }
        }

        int count = 0;

        for (int i = coins.size() - 1; i > -1; i--) {
            if(M == 0) break;
            else if(M < coins.get(i)) continue;
            else{
                count += (M / coins.get(i));
                M %= coins.get(i);
            }
        }

        bw.write(String.valueOf(count));

        bw.flush();
        bw.close();
    }
}

File: JS/greedy/coinZero.js
"use strict";

const fs = require("fs");

const input =
  process.platform === "linux"
    ? fs.readFileSync("/dev/stdin").toString().split("\n")
    : `10 4200
1
5
10
50
100
500
1000
5000
10000
50000`.split("\n");

let [N, K] = input[0].split(" ").map((n) => parseInt(n));
const coins = [];

for (let i = 1; i < N + 1; i++) {
  const coin = parseInt(input[i]);

  coin <= K && coins.push(coin);
}

let count = 0;

for (let i = coins.length - 1; i > -1; i--) {
  if (K === 0) break;
  else if (K < coins[i]) continue;
  else {
    count += parseInt(K / coins[i]);
    K = K % coins[i];
  }
}

console.log(count);
